pkg/gadgets/top/block-io/tracer/tracer.go

```diff
@@ -273,32 +281,65 @@ func (t *Tracer) nextStats() ([]*types.Stats, error) {
 	return stats, nil
 }
 
-func (t *Tracer) run() {
+func (t *Tracer) run(ctx context.Context, cancel context.CancelFunc) {
+	count := t.config.Count
 	ticker := time.NewTicker(t.config.Interval)
 
-	go func() {
-	loop:
-		for {
-			select {
-			case <-t.done:
-				break loop
-			case <-ticker.C:
-				stats, err := t.nextStats()
-				if err != nil {
-					t.eventCallback(&top.Event[types.Stats]{
-						Error: err.Error(),
-					})
-					return
-				}
+	for {
+		select {
+		case <-t.done:
+			// TODO: Once we completely move to use Run instead of NewTracer,
+			// we can remove this as nobody will directly call Stop (cleanup).
+			return
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			stats, err := t.nextStats()
+			if err != nil {
+				t.eventCallback(&top.Event[types.Stats]{
+					Error: fmt.Sprintf("getting next stats: %s", err),
+				})
+				return
+			}
 
-				n := len(stats)
-				if n > t.config.MaxRows {
-					n = t.config.MaxRows
+			n := len(stats)
+			if n > t.config.MaxRows {
+				n = t.config.MaxRows
+			}
+			t.eventCallback(&top.Event[types.Stats]{Stats: stats[:n]})
+
+			if count > 0 {
+				count--
+				if count == 0 {
+					if cancel != nil {
+						cancel()
+					}
+					return
 				}
-				t.eventCallback(&top.Event[types.Stats]{Stats: stats[:n]})
 			}
```

Comment:
I personally find this code a bit difficult to follow (too many conditionals involved). What about using the for loop structure to control it? 

option 1:
```
for i := 0; i < count || count == 0; i++ {
```

option 2: 
```
	if count == 0 {
		count = math.MaxInt
	}

	for i := 0; i < count; i++ {
```

Reply:
I don't know. They both seem also hacky. 

> I personally find this code a bit difficult to follow (too many conditionals involved). 
>
Even now that it is just?

```go
			if count > 0 {
				count--
				if count == 0 {
					return nil
				}
			}
```

Reply:
It's a bit better. Anyway, it's also a personal preference, if you think using those conditionals is better I'm totally fine with this.